src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,248 @@
+  }
+
+  private static String generateWorkerId(SecureRandom random) {
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    return Hex.toString(workerIdBytes);
+  }
+
+  @Override
+  public void start() {
+    start(Executors.newSingleThreadScheduledExecutor());
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executor) {
+    this.executor = executor;
+
+    try {
+      reconciliationCache.cleanUpWorkerSet(WORKER_TTL_MS);
+    } catch (Exception ex) {
+      logger.warn("failed to clean up worker set", ex);
+    }
+
+    this.executor.scheduleWithFixedDelay(new PingTask(workerId, reconciliationCache), 0, WORKER_TTL_MS / 2, TimeUnit.MILLISECONDS);
+    scheduleWithJitter(CHUNK_INTERVAL);
+  }
+
+  @Override
+  public void stop() {
+    this.executor.shutdown();
+
+    try {
+      this.executor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ex) {
+      throw new AssertionError(ex);
+    }
+
+    try {
+      reconciliationCache.leaveWorkerSet(workerId);
+    } catch (Exception ex) {
+      logger.warn("failed to remove from worker set", ex);
+    }
+  }
+
+  @Override
+  public void run() {
+    boolean accelerate  = false;
+    long    workerCount = 1;
+
+    try {
+      if (reconciliationCache.lockActiveWorker(workerId, CHUNK_INTERVAL)) {
+        try {
+          if (processChunk()) {
+            accelerate = reconciliationCache.isAccelerated();
+          }
+        } finally {
+          reconciliationCache.unlockActiveWorker(workerId);
+        }
+      }
+
+      if (!accelerate) {
+        workerCount = Math.max(reconciliationCache.getWorkerCount(WORKER_TTL_MS), 1);
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    if (!accelerate) {
+      scheduleWithJitter(CHUNK_INTERVAL * workerCount);
+    } else {
+      scheduleWithJitter(ACCELERATE_CHUNK_INTERVAL);
+    }
+  }
+
+  private void scheduleWithJitter(long delayMs) {
+    long randomJitterMs = (long) (random.nextDouble() * JITTER_MAX * delayMs);
+    try {
+      executor.schedule(this, delayMs + randomJitterMs, TimeUnit.MILLISECONDS);
+    } catch (RejectedExecutionException ex) {
+      logger.info("reconciler shutting down");
+    }
+  }
+
+  private boolean processChunk() {
+    Optional<String>               fromNumber          = reconciliationCache.getLastNumber();
+    int                            chunkSize           = (int) Math.max(getAccountCount() * PERIOD / CHUNK_INTERVAL, MINIMUM_CHUNK_SIZE);
+    DirectoryReconciliationRequest request             = readChunk(fromNumber, chunkSize);
+    Response                       sendChunkResponse   = sendChunk(request);
+    boolean                        sendChunkSuccessful = sendChunkResponse.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL;
+
+    if (sendChunkResponse.getStatus() == 404 || request.getToNumber() == null) {
```

Comment:
What does 404 signify? Maybe you could have your own DirectoryReconciliationResponse object that helps clarify?

Reply:
yes, will do. 404 made more sense when reconciliation was partition-based.

to clarify, though, this is a way for the contact discovery cluster to signify that it has lost data from earlier in this reconciliation cycle (i.e. catestrophic data loss) and that reconciliation should start from the beginning.

this serves as a decoupling of the concerns "cds needs the whole data set again" and "we want to accelerate reconciliation". i am comfortable with the former being automatically triggered, but making acceleration automatic would make me nervous.

Reply:
see c94595dc2c37. @moxie0 what do you think of using an enum in the json object like that?